### ISSUE TITLE
Determine cause of PHP 8.1 performance regressions

### ISSUE THREAD

**Issue by cmaeng** [department-of-veterans-affairs/va.gov-cms#11469]
## Description
In our efforts to upgrade from PHP 7.4 to 8.1 we did performance testing to compare speeds. #11185 uncovered that 8.1 was noticeably slower than 7.4. We'll want to know why and work to resolve.

## Acceptance Criteria
- [x] The root cause of 8.1 performance regressions are determined
- [x] A ticket is created and refined with steps to resolve

### CMS Team
Please check the team(s) that will do this work.

- [ ] `Program`
- [x] `Platform CMS Team`
- [ ] `Sitewide Crew`
- [ ] `⭐️ Sitewide CMS`
- [ ] `⭐️ Public Websites`
- [ ] `⭐️ Facilities`
- [ ] `⭐️ User support`

**Comment by cmaeng** [comment]
something weird is happening where we're not seeing this in Tugboat. may want to Blackfire this

**Comment by cmaeng** [comment]
based on investigations so far we've determined:
- performance slow downs may not be PHP related as we've seen the same slow down with 7.4
- because the root cause is inconclusive, we would recommend moving forward with putting 8.1 on staging in order to be able to research further

**Comment by cmaeng** [comment]
per @mchelen on 11/15/22:
- if we see content build speeds increased by up to 5 mins, once per day (on cold cache), this is not a launch blocker
- possible editor experience effects need to be run by UX experts within the program to assess

**Comment by ElijahLynn** [comment]
I deployed same codebase on 7.4 and 8.1 on test.staging.cms.va.gov and test.prod.cms.va.gov respectively and overall GQL query times for the content build are near the same, around 9-10 minutes on a cold cache, and then PHP 7.4 with same codebase on staging.cms.va.gov was 5-6 minutes. So I am now working on deploying the 8.1 + codebase branch to the staging.cms.va.gov server to see if the times improve. Maybe there is a networking issue. 

All infra resources appear to be identical, including EC2 instance size/type, RDS instance size type, attached EBS volumes and IOPS, plus memcache clusters. No burst limits appear to have been reached either.

**Comment by ElijahLynn** [comment]
I compared the RDS param groups, and they are identical. 

![image](https://user-images.githubusercontent.com/1504756/202831986-cacd3a19-cf10-40d9-b7a5-3a8a06c9027f.png)

![image](https://user-images.githubusercontent.com/1504756/202832000-94efad67-8a45-413c-8229-c719bed6a8f2.png)

 
![image](https://user-images.githubusercontent.com/1504756/202832044-ab9c9377-d27a-43c5-9864-3c32c43c151c.png)

**Comment by ElijahLynn** [comment]
So, I'm still noticing odd things, but consistent. For instance, we have the same commit deployed to TEST-STAGING and TEST-PROD, both on PHP 8.1, this commit is the latest of our PHP 8.1 branch. Then we have a recent commit on `main` deployed to STAGING.  

The timings for them are consistent and some are showing that our PHP 8.1 code and ENV are twice as slow in some cases on a cold cache. 

![image](https://user-images.githubusercontent.com/1504756/202833415-e4b62bbe-976a-442a-86bb-8f9f7cc77ee3.png)

**Comment by ElijahLynn** [comment]
STAGING is twice as fast as TEST-STAGING and TEST-PROD for /admin/content GET requests. TEST-PROD is consistently significantly slower then STAGING and TEST-STAGING, and TEST-STAGING is consistently slower than STAGING.

**Comment by ElijahLynn** [comment]
`drush cache:rebuild` is roughly the same on STAGING and TEST-STAGING, but TEST-PROD is 30% slower than both. 

![image](https://user-images.githubusercontent.com/1504756/202833604-7349ff3a-9674-4144-adf1-ced9b0c1f3c0.png)

**Comment by cmaeng** [comment]
Next steps:
- Coordinate with our CMS UX folks to determine the threshold by which we are comfortable moving forward or not
- @ElijahLynn to dig in to possible networking issues that might affect this

**Comment by ElijahLynn** [comment]
I've got the test.prod RDS instance moved over to AZ 1C now, so all RDS are consistent in 1C. RDS are multi-AZ fwiw, but I can't actually tell that that is working and only see one instance for each, and they are all in 1C. :shrug:

**Comment by ElijahLynn** [comment]
I have a PR open to lock the instances to 1C for now https://github.com/department-of-veterans-affairs/devops/pull/12276, so they are in the same AZ as RDS. All Memcache clusters have at least 1 node in 1C as well.

**Comment by ElijahLynn** [comment]
Note to self to update docs here with new 81 path > https://github.com/department-of-veterans-affairs/va.gov-cms/blob/main/READMES/drupal-memcache.md

**Comment by ElijahLynn** [comment]
I tested out `zend.assertions = 0` and `assertions.exception = 0` and no noticeable performance gain, https://dsva.slack.com/archives/C02HX4AQZ33/p1669393524296129 for discussion.

**Comment by ElijahLynn** [comment]
I noticed on Wednesday that only one of two memcache nodes were responding to an `nmap` probe, even thought both nodes responsed to a telnet connection. So, I tested out removing the memcache IPs and the cold cache GET to /node/946 went from 14 seconds to 5 seconds. I still need to do more testing around this but I'm thinking maybe we should start with two fresh memcache nodes, or get them both responding or debug why only one is responding.

**Comment by ElijahLynn** [comment]
I need to do further testing but it looks like a cache SET on RDS is much faster than with Memcache (Elasticache). I suppose maybe it is because were are using the `consistent` algorithm, to populate across multiple nodes, and RDS is just going to be faster with that. And that memcache will shine with GETs, although not seeing a big difference on /admin/content for that scenario right now.

**Comment by ElijahLynn** [comment]
So, the memcache nodes are in two different AZs. And sometimes the ideal AZ, 1C, is listed as the second node. And the first node is 1C in other cases. So, I tried swapping the IPs on test.staging, for instance, and not seeing any noticeable differences.

**Comment by ElijahLynn** [comment]
I just made some good headway on one of the issues and that is this. I created two new memcache nodes (nodes 3&4), both in 1C and between variations of 1,2,3 & 4 I have found that the new nodes notably outperform nodes 1&2. I'll post more results soon, but what we have now far outperforms what is on staging (7.1 + codebase). 

I've seen notable improvements on these paths. The 1st time is 8.1 on test.staging with old memcache nodes, and 2nd time is 8.1 on test.staging with new memcache nodes, both in the same AZ. 

* /admin/content (10s > 3s)
* /node/946 (12s > 3s)
* /graphql content-build (8m08s > 4m02s)

I'll post a better matrix tomorrow, but I think we are in a pretty good state to move forward without a performance regression, and in fact a performance improvement as we expected. Fwiw, I also tried rebooting the memcache nodes and even tried one of the old nodes and new nodes in the same AZ and still saw noticeably worse performance. So the end story here is that we need to create new memcache nodes across the entire stack. Cold cache and warm cache times are going to be nearly twice as fast.
